quic/s2n-quic-core/src/datagram/default.rs

```diff
@@ -3,16 +3,161 @@
 
 // s2n-quic's default implementation of the datagram component
 
-use crate::datagram::{Packet, Sender};
+use core::task::{Context, Poll, Waker};
+
+use crate::datagram::{ConnectionInfo, Packet, Sender};
 use alloc::collections::VecDeque;
 use bytes::Bytes;
 
+#[derive(Debug)]
 pub struct DefaultSender {
-    queue: VecDeque<Datagram>,
+    pub queue: VecDeque<Datagram>,
+    capacity: usize,
+    min_packet_space: usize,
+    max_packet_space: usize,
+    smoothed_packet_size: f64,
+    waker: Option<Waker>,
+    max_datagram_payload: u64,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Datagram {
+    pub data: Bytes,
+}
+
+#[non_exhaustive]
+#[derive(Debug, PartialEq)]
+pub enum SendDatagramError {
+    QueueAtCapacity,
+    DataLengthExceedsPeerLimits,
 }
 
-struct Datagram {
-    data: Bytes,
+impl DefaultSender {
+    /// Creates a builder for the default datagram sender
+    pub fn builder() -> Builder {
+        Builder::default()
+    }
+
+    /// Enqueues a datagram for sending it towards the peer.
+    ///
+    /// # Return value
+    ///
+    /// The function returns:
+    ///
+    /// - `Poll::Pending` if the datagram's send buffer capacity is currently exhausted
+    ///   and the datagram was not added to the queue. In this case, the caller should
+    ///   retry sending after the [`Waker`](core::task::Waker) on the provided
+    ///   [`Context`](core::task::Context) is notified.
+    /// - `Poll::Ready(Ok(()))` if the datagram was enqueued for sending.
+    /// - `Poll::Ready(Err(SendDatagramError))` if an error occurred while trying
+    ///   to send the datagram
+    pub fn poll_send_datagram(
+        &mut self,
+        data: bytes::Bytes,
```

Comment:
I would make this take a `&mut bytes::Bytes`, similar to [SendStream::poll_send](https://docs.rs/s2n-quic/latest/s2n_quic/stream/struct.SendStream.html#method.poll_send). You would then replace it with `core::mem::replace(data, Bytes::new())` when you actually accept it. Otherwise, the caller will have to keep cloning the data until it's accepted.

```suggestion
        data: &mut bytes::Bytes,
```

Reply:
🤔 Isn't this technically copying the Bytes? I thought that we were doing a zero-copy thing.

Reply:
Nope. It's replacing the pointer that they provided with an empty Bytes, which doesn't allocate